DROOLS-4689/DROOLS-4690/DROOLS-4691/DROOLS-4692: [DMN Designer] Code Completion - Enable code completion for Literal Expressions

karreiro opened the pull request:
See:
- [DROOLS-4689](https://issues.jboss.org/browse/DROOLS-4689): [DMN Designer] Code Completion - Setup Monaco (loader)
- [DROOLS-4690](https://issues.jboss.org/browse/DROOLS-4690): [DMN Designer] Code Completion - Add Monaco support for 'TextAreaSingletonDOMElementFactory'
- [DROOLS-4691](https://issues.jboss.org/browse/DROOLS-4691): [DMN Designer] Code Completion - Enable code completion for only Literal Expressions
- [DROOLS-4692](https://issues.jboss.org/browse/DROOLS-4692): [DMN Designer] Code Completion - UI adjusts for Monaco component into DMN Boxed Expression component

---

![2019-11-07 14 31 45](https://user-images.githubusercontent.com/1079279/68412704-74bea100-016b-11ea-80b8-c421d9c337c7.gif)

---

Part of an ensemble:
- https://github.com/kiegroup/appformer/pull/836
- https://github.com/kiegroup/kie-wb-common/pull/3000

karreiro commented:
Jenkins, please execute full downstream build.

manstis commented on uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/sys/PatternFlyClientBundle.java:
```diff
@@ -44,4 +44,7 @@
-}
\ No newline at end of file
+
+    @Source("org/uberfire/client/views/static/monaco-editor/dev/vs/loader.js")
```
The MS site suggests to use `dev` for development and `min` for production. 

I don't see any easy way to switch in our build mechanism so wonder if using `min` always would be better?

Our other external libraries use `min` variants.

karreiro replied:
Unfortunately, the `min` version is not working as expected. I reported the issue [here](https://github.com/microsoft/monaco-editor/issues/1608) some days ago. For now, I think we can wait for a response, but if we don't get it fixed from Monaco side, we will need to enable an asset pipeline from our side (we can use this pipeline to minify other assets like `‌daterangepicker.js`). I don't think that it's a good idea to minify Monaco assets from our side now manually, so I'm going to keep this issue on track here: [DROOLS-4734](https://issues.jboss.org/browse/DROOLS-4734).

manstis commented on uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/sys/PatternFlyBootstrapper.java:
```diff
@@ -119,6 +124,28 @@ public static void ensureJQueryUIIsAvailable() {
+
+        return "(new function() { this.require = { baseUrl: '" + baseUrlPath + "' }; "
+                + script
+                + " window.__MONACO_AMD_LOADER__ = _amdLoaderGlobal });";
```
Ouch, there seems to be some hoops you're jumping through to load Monaco..

Is it possible for you to add some JavaDocs [here](https://github.com/kiegroup/appformer/pull/836/files#diff-c5bafbc2a3bfb8f52eaf4f48b434adc1R141) and [here](https://github.com/kiegroup/appformer/pull/836/files#diff-046c0ce38144a43a8111f56cd02d096aR27) explaining the reasons for these _hacks_?

karreiro replied:
I just added the Java Docs. These hacks are needed because the global scope is polluted by other libraries. **Ace** has its own AMD Loader, **Monaco** has its own AMD Loader as well, and there are other libraries relying on functions and objects with generic names in the global scope.

I've tried to update **Ace** with a new version [1] that relies on **RequireJS** as the AMD Loader, so **Monaco** and **Ace** could  use the same loader since both are compatible with **RequireJS**. However, we have other jQuery libraries that rely on functions that invalidate the setup of **RequireJS**. Thus, I gave up on this approach.

Given this scenario, to avoid as many regressions as possible, I'm enclosing **Monaco** instantiation into this temporary scope, and exposing its loader into this namespace: `__MONACO_AMD_LOADER__` - the only global namespace that **Monaco** will use. However, when we require **Monaco** modules, it needs to temporary use the global namespace for calling `define` and `require` functions, that's why I've encapsulated this switch-logic into `MonacoEditorInitializer`.

[1] Another problem with this approach is the number of regressions related to old Ace APIs.

manstis commented on uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/resources/org/uberfire/client/views/PatternFly.gwt.xml:
```diff
@@ -30,4 +30,5 @@
   <stylesheet src="bootstrap-select/css/bootstrap-select.min.css"/>
   <stylesheet src="prettify/bin/prettify.min.css"/>
   <stylesheet src="uberfire-patternfly.css" />
+  <stylesheet src="monaco-editor/dev/vs/editor/editor.main.css"/>
```
If you decide to switch to `min` remember to change this path too!

karreiro replied:
Fixed! It's fine to use the minified CSS :-)

manstis commented on uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/resources/org/uberfire/client/views/PatternFly.gwt.xml:
```diff
@@ -30,4 +30,5 @@
   <stylesheet src="bootstrap-select/css/bootstrap-select.min.css"/>
   <stylesheet src="prettify/bin/prettify.min.css"/>
   <stylesheet src="uberfire-patternfly.css" />
+  <stylesheet src="monaco-editor/min/vs/editor/editor.main.css"/>
```
Did you want to keep this as `dev` given your comment about using the `min` version?

karreiro replied:
I think we're fine regarding CSS compression, Mike :-)

karreiro commented:
Thank you for the review, @manstis. I've applied the changes in a separate commit.

karreiro commented:
@manstis. Thank you for your first review, but I've slightly change this PR with two commits since you approved it.

I got the final list of functions approved by Matteo this morning, so I've updated the list into one commit. And, in another commit, I've fixed the issue related to duplicated suggestions by initialising FEEL language once :-)

karreiro commented:
(rebasing the branch... 👀)

karreiro commented:
Jenkins execute compile downstream build

karreiro commented:
(rebasing the branch... 👀)

jomarko commented:
Done with review, the ensemble **can not be merged**, because it breaks the existing selenium [tests](https://github.com/jboss-integration/bxms-qe-tests/blob/master/test-kie-wb/src/test/java/org/jboss/qa/brms/dmn/DmnTest.java).

manstis commented:
@jomarko Could the test not be ignored short-term? 

A new JIRA for QEto update the test can follow. I suspect there's some research required as to how the automated test can integrate with the Monaco editor used in this PR. We'd ideally not want to stall merge of this feature whilst that effort is undertaken.

> Done with review, the ensemble **can not be merged**, because it breaks the existing selenium [tests](https://github.com/jboss-integration/bxms-qe-tests/blob/master/test-kie-wb/src/test/java/org/jboss/qa/brms/dmn/DmnTest.java).

jomarko commented:
@manstis no, from my point of view, the tests can not be ignored, because the level of uncertainty, if it will be possible to fix them afterward, is too high.

We are in situation merging would break existing tests. I f the merge would affect just possible new tests, I wouldn't block merge.

Trying to update existing ones to the state produced by this PR. No luck till now.

karreiro commented:
(rebasing the branch...)

karreiro commented:
Jenkins, please execute full downstream build.

karreiro commented:
Broken tests on full downstream build are not related to this PR:
```
org.kie.wb.selenium.ui.StandalonePerspectivesIntegrationTest.testLibraryPerspectiveWithHeader
org.kie.wb.selenium.ui.StandalonePerspectivesIntegrationTest.org.kie.wb.selenium.ui.StandalonePerspectivesIntegrationTest
```

karreiro commented:
Jenkins, please execute full downstream build.

jomarko commented:
jenkins execute full downstream build

karreiro commented:
(rebasing the branch...)

karreiro commented:
@jomarko Could please take a look?

jomarko commented on uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/monaco/jsinterop/MonacoLoader.java:
```diff
@@ -0,0 +1,38 @@
+public class MonacoLoader {
+
+    @JsMethod(namespace = GLOBAL, name = "require")
+    public native static void require(final JsArrayString modules,
```
![Screenshot from 2019-12-17 14-16-34](https://user-images.githubusercontent.com/8044780/70998527-db729b00-20d7-11ea-9e4d-94482c43e36e.png)

karreiro replied:
Fixed.

karreiro commented:
(rebasing the branch...)
